I squashed the commits from @qaristote laying in some other branch, and made linking with Z3 conditionally. `liquid-fixpoint` only links Z3 if requested with a cabal flag.

One other change with respect to the previous state of the PR is that the functions to interact asynchronously with the SMT solver have been removed. When measuring them we didn't find any improvements, and they made the code more complicated.

There are some test failures to check in the CI jobs still.

In the job "cabal / ghc-9.2.3 z3-4.10.2"
```
fixpoint: waitForProcess: does not exist ([smtlib-backends-process] while killing the solver process: No child processes)
```

In the circleci job:
```
using typed-process-0.2.8.0 and ghc-8.10.7
...
smtlib-backends-process          > [1 of 1] Compiling SMTLIB.Backends.Process
smtlib-backends-process          > 
smtlib-backends-process          > /tmp/stack-11b3cabb5dc5c281/smtlib-backends-process-0.2/src/SMTLIB/Backends/Process.hs:39:5: error:
smtlib-backends-process          >     Module ‘System.Process.Typed’ does not export ‘mkPipeStreamSpec’
smtlib-backends-process          >    |
smtlib-backends-process          > 39 |     mkPipeStreamSpec,
smtlib-backends-process          >    |     ^^^^^^^^^^^^^^^^
smtlib-backends-process          >
```
One solution is to specify a newer typed-process in the stack configuration, and to adjust the lower bound in the cabal file of `smtlib-backends-process`.

--- a/src/Language/Fixpoint/Smt/Interface.hs
+++ b/src/Language/Fixpoint/Smt/Interface.hs
@@ -91,7 +91,6 @@ import qualified Data.Text.Lazy.Encoding  as LTE
 import qualified Data.Text.Lazy.IO        as LTIO
 import           System.Directory
 import           System.Console.CmdArgs.Verbosity
-import           System.Exit              hiding (die)
 import           System.FilePath
 import           System.IO
 import qualified System.Process.Typed
@@ -309,11 +308,11 @@ makeContext' cfg ctxLog
                   , ctxSymEnv  = mempty
                   }
 
--- | Close file handles and wait for the solver process to terminate.
-cleanupContext :: Context -> IO ExitCode
+-- | Close file handles and release the solver backend's resources.
+cleanupContext :: Context -> IO ()
 cleanupContext Ctx {..} = do
   maybe (return ()) (hCloseMe "ctxLog") ctxLog
-  ctxClose >> return ExitSuccess
+  ctxClose
 
 hCloseMe :: String -> Handle -> IO ()
 hCloseMe msg h = hClose h `catch` (\(exn :: IOException) -> putStrLn $ "OOPS, hClose breaks: " ++ msg ++ show exn)

--- a/src/Language/Fixpoint/Solver/Instantiate.hs
+++ b/src/Language/Fixpoint/Solver/Instantiate.hs
@@ -799,7 +799,7 @@ withCtx cfg file env k = do
   ctx <- SMT.makeContextWithSEnv cfg file env
   _   <- SMT.smtPush ctx
   res <- k ctx
-  _   <- SMT.cleanupContext ctx
+  SMT.cleanupContext ctx
   return res
 
 infixl 9 ~>

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -1094,7 +1094,7 @@ withCtx cfg file env k = do
   ctx <- SMT.makeContextWithSEnv cfg file env
   _   <- SMT.smtPush ctx
   res <- k ctx
-  _   <- SMT.cleanupContext ctx
+  SMT.cleanupContext ctx
   return res
 
 

--- a/stack.yaml
+++ b/stack.yaml
@@ -9,13 +9,19 @@ flags:
 packages:
 - '.'
 
-
 extra-deps:
 - hashable-1.3.5.0
 - rest-rewrite-0.3.0
-- smtlib-backends-0.2
-- smtlib-backends-process-0.2
-- smtlib-backends-z3-0.2
+- git: https://github.com/tweag/smtlib-backends
+  commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+- git: https://github.com/tweag/smtlib-backends
+  commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+  subdir: smtlib-backends-z3
+- git: https://github.com/tweag/smtlib-backends
+  commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+  subdir: smtlib-backends-process
+- typed-process-0.2.10.1@sha256:cc35b76f2ce32c708cb2f68a1c4f661a8138196d958e34a2f214208333a73f46,2070
+
 
 nix:
   packages: [z3]

--- a/stack.yaml.lock
+++ b/stack.yaml.lock
@@ -19,26 +19,49 @@ packages:
   original:
     hackage: rest-rewrite-0.3.0
 - completed:
-    hackage: smtlib-backends-0.2@sha256:8135796ab4cd03c6511dc88d02f152b8dfe32c31d2c31c4867b4d06d03b13899,1205
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
+    name: smtlib-backends
     pantry-tree:
-      sha256: 7d7f39dcf03d13362c9fa7c66a6ad3ef5086d577cc23c8db9cbadb6176f498cb
-      size: 274
+      sha256: 1512451cd1b43d00344113f59a07602db6fadd5bd2e155f15785b9c173b32d61
+      size: 2600
+    version: '0.3'
   original:
-    hackage: smtlib-backends-0.2
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
 - completed:
-    hackage: smtlib-backends-process-0.2@sha256:65bb9c83b39266db0155dce6b5a55a0e9c92cb9c97a8ce70775de0fdac551dea,1762
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
+    name: smtlib-backends-z3
     pantry-tree:
-      sha256: 374a45b6b1cb0f1c81d4505e01057b28d52fe1f70f01c412878e53aa09bbb35d
-      size: 401
+      sha256: 89cd9eed0f80339077a5588c92abe7044f1f259a1e9a82aa7b2f70044553ae44
+      size: 453
+    subdir: smtlib-backends-z3
+    version: '0.3'
   original:
-    hackage: smtlib-backends-process-0.2
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
+    subdir: smtlib-backends-z3
 - completed:
-    hackage: smtlib-backends-z3-0.2@sha256:3656cb3b0a337da9458c499915cd19500abb83257b4fea31e379160148c468c8,1950
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
+    name: smtlib-backends-process
     pantry-tree:
-      sha256: a7f4aace1eafcec5c3bd98eb75c768d56b79e727b6ed231c6ba2ed02f8f7e26b
-      size: 390
+      sha256: 3218f35c149a504fbb5e2eb8a068865ad8a51aec04821b5ed237d3a5c0ca5bb0
+      size: 469
+    subdir: smtlib-backends-process
+    version: '0.3'
   original:
-    hackage: smtlib-backends-z3-0.2
+    commit: 745f24891f0adb065551d2ad9b13dbfb43f504b8
+    git: https://github.com/tweag/smtlib-backends
+    subdir: smtlib-backends-process
+- completed:
+    hackage: typed-process-0.2.10.1@sha256:cc35b76f2ce32c708cb2f68a1c4f661a8138196d958e34a2f214208333a73f46,2070
+    pantry-tree:
+      sha256: b3d6511483d5c869fdfb7220dc13b205730502162a1f6bdca9307a676f3fcb8c
+      size: 529
+  original:
+    hackage: typed-process-0.2.10.1@sha256:cc35b76f2ce32c708cb2f68a1c4f661a8138196d958e34a2f214208333a73f46,2070
 snapshots:
 - completed:
     sha256: 79a786674930a89301b0e908fad2822a48882f3d01486117693c377b8edffdbe

--- a/tests/test.hs
+++ b/tests/test.hs
@@ -7,6 +7,7 @@ module Main where
 import qualified Control.Concurrent.STM as STM
 import qualified Data.Functor.Compose   as Functor
 import qualified Data.IntMap            as IntMap
+import Control.Monad (when)
 import qualified Control.Monad.State    as State
 import Control.Monad.Trans.Class (lift)
 
@@ -134,11 +135,14 @@ mkTest testCmd code dir file
         assertEqual "" True True
       else do
         createDirectoryIfMissing True $ takeDirectory log
-        withFile log WriteMode $ \h -> do
+        c <- withFile log WriteMode $ \h -> do
           let cmd     = testCmd opts "fixpoint" dir file
           (_,_,_,ph) <- createProcess $ (shell cmd) {std_out = UseHandle h, std_err = UseHandle h}
-          c          <- waitForProcess ph
-          assertEqual "Wrong exit code" code c
+          waitForProcess ph
+        when (code /= c) $
+          readFile log >>= putStrLn
+        assertEqual "Wrong exit code" code c
+
   where
     test = dir </> file
     log  = let (d,f) = splitFileName file in dir </> d </> ".liquid" </> f <.> "log"

